File: scripts/import-review/ready_to_publish_gate.py
"""Quality gate for import_review_items.review_status = ready_to_publish.

Mirror of lib/import-review/ready-to-publish-gate.ts — keep the rules in sync.
"""

from __future__ import annotations

import re
from typing import Any

READY_DUPLICATE_BLOCKLIST = frozenset(
    {"recurring_ad", "exact_duplicate", "likely_duplicate"}
)
LOCKED_STATUSES = frozenset({"approved", "rejected", "duplicate", "quarantine"})

HANDLE_TOKEN_RE = re.compile(
    r"\b@?[A-Za-z][A-Za-z0-9]*(?:_[A-Za-z0-9]+){1,}\b"
)
BARE_USERNAME_RE = re.compile(r"^@?[A-Za-z][A-Za-z0-9._]{2,31}$")
LETTER_RE = re.compile(r"[A-Za-zА-Яа-яЁё]")

JUNK_TITLES = {
    "messenger",
    "whatsapp",
    "telegram",
    "gmail.com",
    "yahoo.com",
    "mail.com",
    "instagram",
    "facebook",
    "unknown",
    "user",
    "admin",
    "null",
    "none",
    "n/a",
    "без названия",
}


def _first_filled(*values: Any) -> str:
    for raw in values:
        t = re.sub(r"\s+", " ", str(raw or "")).strip()
        if t:
            return t
    return ""


def _has_nonempty(value: Any) -> bool:
    if isinstance(value, list):
        return any(str(x or "").strip() for x in value)
    return bool(str(value or "").strip())


def display_title(row: dict[str, Any]) -> str:
    return _first_filled(
        row.get("title"), row.get("business_name"), row.get("person_name")
    )


def is_unusable_ready_title(
    raw: str | None,
    *,
    description: str | None = None,
    source_text: str | None = None,
) -> bool:
    t = re.sub(r"\s+", " ", raw or "").strip()
    if not t or len(t) <= 1:
        return True
    if len(LETTER_RE.findall(t)) < 3:
        return True
    if t.lower() in JUNK_TITLES:
        return True
    if "@" in t:
        return True
    if BARE_USERNAME_RE.match(t) and ("." in t or "_" in t):
        return True
    if HANDLE_TOKEN_RE.search(t) and len(t.split()) <= 5:
        return True
    if len(t) > 90:
        return True
    blob = _first_filled(description, source_text)
    if blob and len(t) >= 48:
        compact = t[:80]
        if blob.startswith(compact) or compact in blob:
            return True
    return False


def qualifies_ready_to_publish(row: dict[str, Any]) -> tuple[bool, str | None]:
    has_phone = _has_nonempty(row.get("phone"))
    has_city = bool(str(row.get("city") or "").strip())
    if not has_phone and not has_city:
        return False, "no_phone_or_city"

    dup = str(row.get("duplicate_status") or "").strip().lower()
    if dup in READY_DUPLICATE_BLOCKLIST:
        return False, "duplicate"

    title = display_title(row)
    if is_unusable_ready_title(
        title,
        description=row.get("description"),
        source_text=row.get("source_text"),
    ):
        return False, "unusable_title"
    return True, None


def status_after_ready_gate(
    row: dict[str, Any],
    requested: str | None = None,
    *,
    prefer_ready: bool = False,
) -> str:
    wanted = (requested or row.get("review_status") or "pending" or "pending")
    wanted = str(wanted).strip() or "pending"
    if wanted in LOCKED_STATUSES:
        return wanted

    wants_ready = wanted == "ready_to_publish" or prefer_ready
    if not wants_ready:
        return wanted

    ok, reason = qualifies_ready_to_publish(row)
    if ok:
        return "ready_to_publish"
    if reason == "no_phone_or_city":
        return "needs_more_info"
    return "pending"

File: lib/import-review/ready-to-publish-gate.ts
/**
 * Quality gate for import_review_items.review_status = ready_to_publish.
 * Shared by Inbox, admin promote, and ingest (TS). Python mirror:
 * scripts/import-review/ready_to_publish_gate.py
 */

import { isJunkImportTitle } from "@/lib/import-review/display-name";

export const READY_DUPLICATE_BLOCKLIST = [
  "recurring_ad",
  "exact_duplicate",
  "likely_duplicate",
] as const;

export type ReadyToPublishFailReason =
  | "no_phone_or_city"
  | "unusable_title"
  | "duplicate";

export type ReadyToPublishInput = {
  title?: string | null;
  business_name?: string | null;
  person_name?: string | null;
  description?: string | null;
  source_text?: string | null;
  phone?: string | string[] | null;
  city?: string | null;
  duplicate_status?: string | null;
};

const HANDLE_TOKEN_RE = /\b@?[a-z][a-z0-9]*(?:_[a-z0-9]+){1,}\b/i;
const BARE_USERNAME_RE = /^@?[A-Za-z][A-Za-z0-9._]{2,31}$/;
const LETTER_RE = /\p{L}/gu;

function letterCount(value: string): number {
  return (value.match(LETTER_RE) || []).length;
}

function firstFilled(values: Array<string | null | undefined>): string {
  for (const raw of values) {
    const t = (raw || "").replace(/\s+/g, " ").trim();
    if (t) return t;
  }
  return "";
}

function hasNonEmpty(value: string | string[] | null | undefined): boolean {
  if (Array.isArray(value)) {
    return value.some((x) => String(x || "").trim().length > 0);
  }
  return Boolean(String(value || "").trim());
}

/** Telegram handle, one character, empty, or the whole post pasted as the name. */
export function isUnusableReadyTitle(
  raw: string | null | undefined,
  extras?: { description?: string | null; sourceText?: string | null },
): boolean {
  const t = (raw || "").replace(/\s+/g, " ").trim();
  if (!t) return true;
  if (t.length <= 1) return true;
  if (letterCount(t) < 3) return true;
  if (isJunkImportTitle(t)) return true;
  if (BARE_USERNAME_RE.test(t) && /[._]/.test(t)) return true;
  if (HANDLE_TOKEN_RE.test(t) && t.split(" ").length <= 5) return true;
  if (t.length > 90) return true;

  const blob = firstFilled([extras?.description, extras?.sourceText]);
  if (blob && t.length >= 48) {
    const compactTitle = t.slice(0, 80);
    if (blob.startsWith(compactTitle) || blob.includes(compactTitle)) {
      return true;
    }
  }
  return false;
}

export function displayTitleForReadyGate(input: ReadyToPublishInput): string {
  return firstFilled([input.title, input.business_name, input.person_name]);
}

export function qualifiesReadyToPublish(input: ReadyToPublishInput): {
  ok: boolean;
  reason?: ReadyToPublishFailReason;
} {
  const hasPhone = hasNonEmpty(input.phone);
  const hasCity = Boolean((input.city || "").trim());
  if (!hasPhone && !hasCity) {
    return { ok: false, reason: "no_phone_or_city" };
  }

  const dup = (input.duplicate_status || "").trim().toLowerCase();
  if (
    READY_DUPLICATE_BLOCKLIST.includes(
      dup as (typeof READY_DUPLICATE_BLOCKLIST)[number],
    )
  ) {
    return { ok: false, reason: "duplicate" };
  }

  const title = displayTitleForReadyGate(input);
  if (
    isUnusableReadyTitle(title, {
      description: input.description,
      sourceText: input.source_text,
    })
  ) {
    return { ok: false, reason: "unusable_title" };
  }

  return { ok: true };
}

const LOCKED_STATUSES = new Set([
  "approved",
  "rejected",
  "duplicate",
  "quarantine",
]);

/**
 * Apply the ready gate to a requested queue status. Never publishes.
 * `preferReady` is for AI-accepted ingest that would otherwise stay pending.
 */
export function statusAfterReadyGate(
  input: ReadyToPublishInput,
  requested: string | null | undefined,
  opts?: { preferReady?: boolean },
): {
  status: string;
  reason?: ReadyToPublishFailReason;
} {
  const wanted = (requested || "pending").trim() || "pending";
  if (LOCKED_STATUSES.has(wanted)) {
    return { status: wanted };
  }

  const wantsReady = wanted === "ready_to_publish" || Boolean(opts?.preferReady);
  if (!wantsReady) {
    return { status: wanted };
  }

  const result = qualifiesReadyToPublish(input);
  if (result.ok) {
    return { status: "ready_to_publish" };
  }
  if (result.reason === "no_phone_or_city") {
    return { status: "needs_more_info", reason: result.reason };
  }
  return { status: "pending", reason: result.reason };
}
